pkg/controllers/workspace_controller.go

```diff
@@ -274,75 +261,56 @@ func (c *WorkspaceReconciler) updateControllerRevision(ctx context.Context, wObj
 
-	if annotations == nil {
-		annotations = make(map[string]string)
-	} // nil checking.
-
-	annotations[WorkspaceRevisionAnnotation] = currentHash
+	annotations[WorkspaceHashAnnotation] = currentHash
 	wObj.SetAnnotations(annotations)
-	deployment := &appsv1.Deployment{}
-	if wObj.Inference != nil {
-		if previousWObj == nil || !compareAdapters(previousWObj.Inference.Adapters, wObj.Inference.Adapters) {
-			if err := c.Get(ctx, types.NamespacedName{
-				Name:      wObj.Name,
-				Namespace: wObj.Namespace,
-			}, deployment); err != nil {
-				if !errors.IsNotFound(err) {
-					klog.ErrorS(err, "failed to get deployment", "deplyment", wObj.Name)
-				}
-				return client.IgnoreNotFound(err)
-			}
-			if deployment.Annotations == nil {
-				deployment.Annotations = make(map[string]string)
+	controllerRevision := &appsv1.ControllerRevision{}
+	if err := c.Get(ctx, types.NamespacedName{
+		Name:      newRevision.Name,
+		Namespace: newRevision.Namespace,
+	}, controllerRevision); err != nil {
+		if errors.IsNotFound(err) {
+			if err := c.Create(ctx, newRevision); err != nil {
+				return fmt.Errorf("failed to create new ControllerRevision: %w", err)
+			} else {
+				annotations[kaitov1alpha1.WorkspaceRevisionAnnotation] = strconv.FormatInt(revisionNum, 10)
 			}
 
-			if hash, exists := deployment.Annotations[WorkspaceRevisionAnnotation]; !exists || (hash != currentHash) {
-
-				var volumes []corev1.Volume
-				var volumeMounts []corev1.VolumeMount
-				shmVolume, shmVolumeMount := utils.ConfigSHMVolume(*wObj.Resource.Count)
-				if shmVolume.Name != "" {
-					volumes = append(volumes, shmVolume)
-				}
-				if shmVolumeMount.Name != "" {
-					volumeMounts = append(volumeMounts, shmVolumeMount)
-				}
-
-				if len(wObj.Inference.Adapters) > 0 {
-					adapterVolume, adapterVolumeMount := utils.ConfigAdapterVolume()
-					volumes = append(volumes, adapterVolume)
-					volumeMounts = append(volumeMounts, adapterVolumeMount)
-				}
-
-				initContainers, envs := resources.GenerateInitContainers(wObj, volumeMounts)
-				spec := &deployment.Spec
-				spec.Template.Spec.InitContainers = initContainers
-				spec.Template.Spec.Containers[0].Env = envs
-				spec.Template.Spec.Containers[0].VolumeMounts = volumeMounts
-				deployment.Annotations[WorkspaceRevisionAnnotation] = currentHash
-				spec.Template.Spec.Volumes = volumes
-
-				if err := c.Update(ctx, deployment); err != nil {
-					return fmt.Errorf("failed to update deployment: %w", err)
+			if len(revisions.Items) > consts.MaxRevisionHistoryLimit {
+				if err := c.Delete(ctx, &revisions.Items[0]); err != nil {
+					return fmt.Errorf("failed to delete old revision: %w", err)
 				}
 			}
-
+		} else {
+			return fmt.Errorf("failed to get controller revision: %w", err)
+		}
+	} else {
+		if controllerRevision.Annotations[WorkspaceHashAnnotation] != newRevision.Annotations[WorkspaceHashAnnotation] {
+			return fmt.Errorf("revision has the same name and namespace but different workspaces")
```

Comment:
"revision has the same name and namespace but different workspaces" -> "revision name conflicts, the hash values are different"